store/src/java/com/zimbra/cs/service/mail/FileSharedWithMe.java

```diff
@@ -0,0 +1,84 @@
+
+import java.util.Map;
+
+import com.zimbra.common.mailbox.FolderConstants;
+import com.zimbra.common.service.ServiceException;
+import com.zimbra.common.soap.Element;
+import com.zimbra.common.soap.MailConstants;
+import com.zimbra.cs.account.Account;
+import com.zimbra.cs.account.Provisioning;
+import com.zimbra.cs.mailbox.Mailbox;
+import com.zimbra.cs.mailbox.MailboxManager;
+import com.zimbra.cs.mailbox.OperationContext;
+import com.zimbra.soap.ZimbraSoapContext;
+
+public class FileSharedWithMe extends MailDocumentHandler implements DelegatableRequest {
+
+    private static final String REVOKE = "revoke";
+    private static final String EDIT = "edit";
+    private static final String CREATE = "create";
+
+    public Element handle(Element request, Map<String, Object> context) throws ServiceException {
+        try {
+            ZimbraSoapContext zsc = getZimbraSoapContext(context);
+            String ownerAccountId = zsc.getAuthtokenAccountId();
+            String granteeAccountID = zsc.getRequestedAccountId();
+            OperationContext octxt = getOperationContext(zsc, context);
+            Account account = getRequestedAccount(zsc);
+            Mailbox mbox = MailboxManager.getInstance().getMailboxByAccount(account, false);
+
+            String action = request.getAttribute(MailConstants.E_ACTION);
+            String fileName = request.getAttribute(MailConstants.A_CONTENT_FILENAME);
+            int ownerFileId = request.getAttributeInt(MailConstants.A_ITEMID);
+            String fileUUID = request.getAttribute(MailConstants.A_REMOTE_UUID);
+            String fileOwnerName = request.getAttribute(MailConstants.A_OWNER_NAME);
+            String rights = request.getAttribute(MailConstants.A_RIGHTS);
+            String contentType = request.getAttribute(MailConstants.A_CONTENT_TYPE);
+            long size = request.getAttributeLong(MailConstants.A_SIZE);
+            if (Provisioning.onLocalServer(account)) {
+                // if file is shared, this file should be available at receivers end
+                if (action.equals(CREATE)) {
+                    mbox.createFileSharedWithMe(mbox, octxt, FolderConstants.ID_FOLDER_FILE_SHARED_WITH_ME, fileName,
+                            ownerAccountId, ownerFileId, fileUUID, fileOwnerName, contentType, size, rights);
+                } // if file access is revoked at source, delete it from receiver end as well
+                else if (action != null && action.equals(REVOKE)) {
+                    mbox.deleteFileSharedWithMe(mbox, octxt, fileUUID, ownerFileId, ownerAccountId, granteeAccountID);
+                } // if file rights is changed, update file permission accordingly
+                else if (action != null && action.equalsIgnoreCase(EDIT)) {
+                    mbox.updateFileSharedWithMe(mbox, octxt, ownerAccountId, ownerFileId, fileUUID, fileOwnerName,
+                            contentType, rights, granteeAccountID);
+                }
```

Comment:
@k-kato in this API `ownerAccountId` and `granteeAccountID`  are retrieved from `zsc` . this is possible to only retrieved when the call is from UI. when the request is from SOAP UI the auth token would contain the same `ownerAccountId` and `granteeAccountID`, added an extra check if it is the same then don't share the file as sharing the file to self is not a valid request. for other validations like invalid Id, or invalid UUID it would be taken care of in DB operation. The reason of extracting `ownerAccountId` and `granteeAccountID` from zcs was to implement security concern so that no 3rd party could make delete/update operation

Reply:
I understand about values retrieved from `zsc`. However, for other variables/parameters, is it okay not to check values?
What will happen when
* fileName, contentType or size does not match the actual file's one?
* ownerFileId or fileUUID is a non-existing ID or invalid format?
* fileOwnerName is a strange name? Does it need to match actual owner name?
* rights is an invalid value like "abc"?

I don't think they are checked in Mailbox.java#createFileSharedWithMe around line 9086...

> it would be taken care of in DB operation

`new DbMailItem(mbox).create(data);` does not check the above but just writes an entry using given values.

Reply:
As we discussed, FileSharedWithMeRequest is called only from zimbra server via proxyRequest in SendShareNotification class. It cannot be called directly from outside of zimbra. Then all parameters in a soap request are always valid.

Reply:
> It cannot be called directly from outside of zimbra.

It's part of the API so it can be. https://github.com/Zimbra/zm-mailbox/pull/1326/files#diff-aab6f1809776ef672dc9da1243f8c875a1d496cdad83b518f0aa3d35dbbe59a5R260

Reply:
> It cannot be called directly from outside of zimbra.

It was not exact. 
FileSharedWithMeRequest can be called from anywhere. However, as per previous discussion, ownerAccountId and granteeAccountID will be the same when it is called directly (= without using proxyRequest). Then it is addressed as an error in the code below. 
https://github.com/Zimbra/zm-mailbox/pull/1326/files#diff-48c461d013fb1197ea91e69b6c526f9f48088974958f12df74dd9512d20dccceR44
```
            if (ownerAccountId.equals(granteeAccountID)) {
                ZimbraLog.mailbox.error("File sharer and grantee cannot be same", new Throwable("invalid request source"));
                throw ServiceException.FAILURE("invalid request", new Throwable("invalid"));
            }
```
It will be described in API reference.
https://github.com/Zimbra/zm-mailbox/pull/1326/files#diff-f9dde6e255750d299d6da695c0170bc0fc1537a8c011382b2e3f4f676749c23bR32
```
This is an internal API, cannot be invoked directly
```